src/main/software/amazon/event/ruler/ComparableNumber.java

```diff
@@ -3,104 +3,122 @@
-        }
-
-        final long shiftedBySixDecimals = number.multiply(TEN_E_SIX).longValueExact();
+        final BigDecimal bigDecimal = JavaBigDecimalParser.parseBigDecimal(str);
+        final double doubleValue = bigDecimal.doubleValue();
 
-        // faster than doing bigDecimal comparisons
-        if (shiftedBySixDecimals < -HALF_TRILLION_TEN_E_SIX || shiftedBySixDecimals > HALF_TRILLION_TEN_E_SIX) {
-            throw new IllegalArgumentException("Value must be between " + -ComparableNumber.HALF_TRILLION +
-                    " and " + ComparableNumber.HALF_TRILLION + ", inclusive");
+        // make sure we have the comparable numbers and haven't eaten up decimals values
+        if(Double.isNaN(doubleValue) || Double.isInfinite(doubleValue) ||
+                BigDecimal.valueOf(doubleValue).compareTo(bigDecimal) != 0) {
+            throw new IllegalArgumentException("Cannot compare number : " + str);
         }
+        final long bits = Double.doubleToRawLongBits(doubleValue);
 
-        return longToBase64Bytes(shiftedBySixDecimals + HALF_TRILLION_TEN_E_SIX);
+        // if high bit is 0, we want to xor with sign bit 1 << 63, else negate (xor with ^0). Meaning,
+        // bits >= 0, mask = 1000000000000000000000000000000000000000000000000000000000000000
+        // bits < 0,  mask = 1111111111111111111111111111111111111111111111111111111111111111
+        final  long mask = ((bits >>> 63) * 0xFFFFFFFFFFFFFFFFL) | (1L  << 63);
+        return numbits(bits ^ mask );
     }
 
-    public static String longToBase64Bytes(long value) {
-        if (value < 0) {
-            throw new IllegalArgumentException("Input value must be non-negative");
+    /**
+     * Converts a long value to a Base128 encoded string representation.
+     * <br/>
+     * The Base128 encoding scheme is a way to represent a long value as a sequence
+     * of bytes, where each byte encodes 7 bits of the original value. This allows for
+     * efficient storage and transmission of large numbers.
+     * <br/>
+     * The method first determines the number of trailing zero bytes in the input
+     * value by iterating over the bytes from the most significant byte to the least
+     * significant byte, and counting the number of consecutive zero bytes at the end.
+     * It then creates a byte array of fixed length {@code MAX_LENGTH_IN_BYTES} and
+     * populates it with the Base128 encoded bytes of the input value, starting from
+     * the least significant byte.
+     * <br/>
+     * As shown in Quamina's numbits.go, it's possible to use variable length encoding
+     * to reduce storage for simple (often common) numbers but it's not done here to
+     * keep range comparisons simple for now.
+     *
+     * @param value the long value to be converted
+     * @return the Base128 encoded string representation of the input value
+     */
+    public static String numbits(long value) {
```

Comment:
Each char in a java String is 2 bytes because they are UTF-16 codepoints. You could save a lot of memory if you could work on the UTF-8 bytes. But I realize that this comment applies to all of Ruler I guess.

Reply:
that's fair. It would be worth a look when we dig into memory & cpu improvements down the line.

Reply:
How are Strings-by-chars sorted in Event Ruler? Binary value or with some collation? If it's binary... you can still use the same scheme but store 2 7-bit bytes per char. Might require shuffling the bytes around if the endianness of the char 2-byte-tuples is different. But as long as leaving the 0x8080 bits out stays clear of combiners or possibly causing illegal codepoints... you should be good.

Reply:
... or use the lower 15 bits. simpler than a 0x8080 mask. Still, check that you get by with only one codepoint and that it's not reserved or illegal in UTF16

Reply:
Comparing ranges got trickier with using 2 7-bit bytes per char, so I've parked the idea for now. Will come back to it later. 

In case anyone takes a stab at it in future, here's the reference code for this : 

```
static final int MAX_LENGTH_IN_BYTES = 5;
...
public static String numbits(long value) {
        char[] result = new char[MAX_LENGTH_IN_BYTES];
        int index = MAX_LENGTH_IN_BYTES - 1;

        while (value != 0 && index >= 0) {
            int lowBits = (int)(value & BASE_128_BITMASK);
            value >>= 7;
            int highBits = (int)(value & BASE_128_BITMASK);
            value >>= 7;

            result[index--] = (char)((highBits << 7) | lowBits);
        }

        return new String(result);
    }
```

Reply:
... Tim told me quamina internally works different. Quamina needs the UTF8 validity because it uses UTF8-illegal byte values as markers. 
Is ruler different enough so you don't depend on that? In that case, you could also just cut it into 4 different shorts. Unless that destroys some string operations you need, e.g by producing hanging combiners etc.

Reply:
For range comparison, Ruler has a lookup table that changes incoming numbers into a sequence of digits https://github.com/aws/event-ruler/blob/main/src/main/software/amazon/event/ruler/Range.java#L128 . Changing this requires more effort and I suspect not as beneficial compared to supporting variable length for numbers. So I chose to work on variable length first and then come back to compressing the numbers later (though I'm hopeful that I'd be able to do both once I get more into refactoring this part of the code more)

Reply:
> Quamina needs the UTF8 validity because it uses UTF8-illegal byte values as markers.
> Is ruler different enough so you don't depend on that

I don't know the answer for this on the top of mind. Will check on it later but I was originally suspecting it should be fine as long as I address any edge cases within the look-up table part of the code.